### ISSUE TITLE
Hits missed when clustering or searching with short sequences

### ISSUE THREAD

**Issue by torognes** [torognes/vsearch#328]
When clustering or searching with short sequences, obvious hits may be missed. This is also a problem when much of the sequences are masked. It is probably due to few distinct unmasked k-mers in the sequences and the required minimum number of shared k-mers (12, set by the `--minwordmatches` option). These heuristics may need to be tuned to work better in these cases.

See also [a VSEARCH Forum post](https://groups.google.com/forum/?#!msg/vsearch-forum/1DvtTJ3DX5I/gSjI_w2YDQAJ) where this issue was described.

**Comment by JDavidson2019** [comment]
Hello, @torognes ,

Has there been any update on if the underlying cause of these missed hits? I am currently attempting to search for short sequences (which are primers < 25 bp) in inverted repeat regions which can be interspersed in regions up to 7 kbp. 

I have attached (in a zip file) a smaller example of trying to search for a primer (`testPrimer.fasta`) in a 2 kbp sequence (`testSeq.fasta`) which has exact matches in two places (a quick ctrl+F with the sequences will return the exact location). I have attached the output (`testVSEARCHOut.aln`) The actual data I am using has about 150,000 reads which have similar inverted repeat regions. The output, however, only results in one match, not two. I have observed similar behavior when the search sequence is extended to 7 kbp with about 10 occurrences of the query sequence.

Unfortunately, even when I set `--minwordmatches 0` to bypass this issue as you suggested in the forum link, the hits are still missed. I have tried various combinations of masking options on both the database and query sequence to no avail. Are there any other fixes that you may think of that will allow sequences such as these to have hits returned? I will also be looking into another tool you developed, `swipe` to see if that may be a better fit for this use case.

Thanks in advance!
[testVsearch.zip](https://github.com/torognes/vsearch/files/12023481/testVsearch.zip)

**Comment by torognes** [comment]
Thanks for your detailed comment. Unfortunately there has not been any changes to this in vsearch lately. I hope to get time to look into it when back from vacation.

**Comment by torognes** [comment]
I've now looked closer at your example. In this case vsearch will not find the second match simply because vsearch will never report more than one match in each database sequence, unless they are on different strands. The program is simply not designed to do that.

**Comment by frederic-mahe** [comment]
Yes, vsearch performs (semi)-global pairwise alignments ([Needleman–Wunsch](https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm)), not local pairwise alignments ([Smith–Waterman](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm)).

> vsearch will never report more than one match in each database sequence, unless they are on different strands.

Tests covering that assertion have been added to our test suite https://github.com/frederic-mahe/vsearch-tests/commit/e256e9df7d1d412d59c7f083ef817bdc247deaa3

**Comment by frederic-mahe** [comment]
There are three problems behind this issue:
1. **the requirement ignored the target's k-mer count**: a target with fewer than `--minwordmatches` distinct k-mers was unreachable, even by  an exact match. Fixed in 919ac670 (see below for more details)
2. **a short query must match all of its words**: under 19 nucleotides at the defaults, one mismatch loses the hit. *Not fixed.*
3. **the k-mer filtering threshold is a constant** — short but k-mer-rich pairs share too few words even at high identity. *Not fixed.*

The candidate-selection threshold was `min(--minwordmatches, distinct k-mers in the query)` — the query's count only. But a target cannot share more k-mers than it holds, so a target with fewer than `--minwordmatches` distinct k-mers was asked for more shared k-mers than arithmetically possible, and could not be selected by any query, not even by an exact match. With the default `--wordlength 8` and `--minwordmatches 12` that was every database sequence with less than `8 + 12 - 1 = 19` nt of distinct k-mer content, and the same pair gave opposite answers depending on direction:
  
| | before | after |
|---|---|---|
| exact 18-nt sequence as query, 400-nt database | 1 hit | 1 hit |
| 400-nt query, exact 18-nt database sequence | **0 hits** | 1 hit |
| 30-nt `(AC)x15` database sequence (2 distinct 8-mers), query containing it verbatim | **0 hits** | 1 hit |
 
The threshold is now capped by the target's own count as well, so a short or heavily masked sequence gives the same answer in either direction. It can only ever lower a threshold that was unreachable, so no pair that was findable before changes, and it measures perf-neutral on `usearch_global` and `cluster_fast`.

Low-complexity clustering benefits too: the two 51-nt sequences of https://groups.google.com/g/vsearch-forum/c/1DvtTJ3DX5I now cluster together with default settings.

**What is not fixed** — the other two parts of the diagnosis in the issue body are untouched, so I am leaving this issue open.

- a query holding fewer distinct k-mers than `--minwordmatches` must still match *all* of them, so a single mismatch loses the hit (an 18-nt query at 94.4% identity: missed);
- the threshold is still a constant, so short but k-mer-rich pairs are still filtered out. A 25-bp primer with one mismatch (96% identity) against a 525-nt template shares 10 k-mers where 12 are required and is still missed, and 40-nt reads at ≥95% identity still over-split when clustered at `--id 0.90`.

Both are tuning questions rather than bugs: they need a decision on whether the default should become length- and identity-aware, which changes results and needs benchmarking across the search commands. Until then the workaround for short or heavily masked sequences is `--minwordmatches 1` (as sensitive as `0` and much faster) or a shorter `--wordlength`. The manual entries for `--minwordmatches`, `--id` and `--dbmask` now state the rule and the workaround, which they did not before.
